New feature: Differentiate between user roles

lesser1 opened the pull request:
- Add dialog to switch between user roles available to the user
- Store (in the pinia store) user roles of current user and the active user role
- Get user roles from initialized Keycloak instance right after the user logged in
- Modify existing UI components to display UI parts depending on the active role of the user
- Add initial navigation safeguards
- Add access rights (whitelists for the main user roles)

[The Gropius Issue](https://frontend.gropius.duckdns.org/components/9d12d6ac-e85a-407c-bdff-a8321fb75d3a/issues/f71fffc8-10b6-408a-ad25-2c213f53f593)

### Definition of Done
- [x] Requirements of the issue are met
- [x] Code has been reviewed
- [x] Code is documented

lesser1 committed:
New feature: Differentiate between user roles

- Add dialog to switch between user roles available to the user
- Store (in the pinia store) user roles of current user and the
  active user role
- Get user roles from initialized Keycloak instance right after the
  user logged in
- Modify existing UI components to display UI parts depending on the
  active role of the user
- Add initial navigation safeguards
- Add access rights (whitelists for the main user roles)

lesser1 committed:
Lint and Format (forgot to do that...)

lesser1 committed:
Fix error caused by referencing deleted getters() from the store

nk-coding commented on src/App.vue:
```diff
@@ -6,5 +6,7 @@
 const store = useAppStore()
-store.initLogin()
+
+const logToken = process.env.NODE_ENV === 'development'
```
I'm not sure if we are there yet.
Because the docker deployment is in production
And then I have no way to get a token currently, which would make testing very very difficult.
Overall good idea, but let's maybe wait until we have a gateway with token in the frontend

lesser1 replied:
Alright, I will change that!

nk-coding commented on src/components/SwitchUserRoleDialog.vue:
```diff
@@ -0,0 +1,82 @@
+<template>
+    <v-dialog max-width="504" persistent>
```
this looks like a very random number
maybe change to 500?

lesser1 replied:
Idk exactly why I am doing it this way, but I always choose numbers that are a multiple of 24... 480 + 24 = 504. However, I see what you mean -- one sees that number and asks themselves 'Hey, why 504???'. I can change it of course.

lesser1 replied:
I thought about this and I've decided to change it to 500 -- this way it's a clean and not-so-random number, I like it.

nk-coding commented on src/components/SwitchUserRoleDialog.vue:
```diff
@@ -0,0 +1,82 @@
+            <v-card-item>
+                <v-card-title>Switch User Role</v-card-title>
+                <v-card-subtitle
+                    >TODO Explain the purpose of this dialog to the
```
TODO?

lesser1 replied:
Well... good thing we do these code reviews... xD

nk-coding commented on src/components/SwitchUserRoleDialog.vue:
```diff
@@ -0,0 +1,82 @@
+ */
+const emit = defineEmits<{
+    (event: 'close'): void
+    (event: 'close-and-navigate-to-storefront'): void
```
why not also use a boolean parameter here like you did in the parent class (to indicate whether you should navigate to storefront)

lesser1 replied:
No particular reason why not, I really like your idea, I will change it.

nk-coding commented on src/store/app.ts:
```diff
@@ -27,30 +39,91 @@ export const useAppStore = defineStore('app', {
+                    )
+                }
+
+                try {
```
I think you could extract some stuff here into functions
this is a bit long

lesser1 replied:
I'm on it, thumbs up.

nk-coding commented on src/store/app.ts:
```diff
@@ -27,30 +39,91 @@ export const useAppStore = defineStore('app', {
+                            return
+                        } else {
+                            const additionalUserRoles: UserRole[] = []
+                            if (
```
looks a bit duplicated, maybe deduplicate with a for loop?

lesser1 replied:
I will consider doing so.

nk-coding commented on src/store/app.ts:
```diff
@@ -5,11 +5,23 @@ import Keycloak from 'keycloak-js'
+export enum UserRole {
+    Buyer = 'Buyer',
+    Employee = 'Employee',
+    SiteAdmin = 'Administrator',
```
the role in only named admin

lesser1 replied:
But the Keycloak role name is "site-admin". Btw, in our domain model it says "Customer", "Employee", and "Admin". Maybe we should decide on one set of names.

lesser1 replied:
I settled on: The frontend defines the roles from the domain model and has a lookup table for mapping the user roles to the role names coming from Keycloak.

nk-coding replied:
well, but the domain model roals will mirror keycloak roles as soon as I fix the domain model (today or tomorrow)
I would really recommend already using the keycloak names

nk-coding reviewed (commented):
This is only code yet, I will test the stuff tomorrow

nk-coding commented on src/store/app.ts:
```diff
@@ -27,30 +39,91 @@ export const useAppStore = defineStore('app', {
+                                ...additionalUserRoles
+                            )
+
+                            console.log(this.userRolesOfCurrentUser)
```
this logs a ref which contains a list
the output is completely unreadable
see https://vuejs.org/api/reactivity-advanced.html#toraw

nk-coding reviewed (changes requested):
When I am not logged in and navigate to `http://localhost:4000/manage-tax-rates`, I just get a white screen, no redirect, no error, no nothing
Same happens when I'm logged in as a buyer
I don't think this is desired behavior

lesser1 commented:
> When I am not logged in and navigate to `http://localhost:4000/manage-tax-rates`, I just get a white screen, no redirect, no error, no nothing Same happens when I'm logged in as a buyer I don't think this is desired behavior

Yes, I noticed that, too. However, I decided to leave that for a dedicated issue regarding redirects and navigation safeguards. For this issue, I only wanted to enforce that one cannot see the contents of a page when they don't have the needed access rights. I want to keep the scope of these PRs relatively small and compact and focused on the original scope of the issue. Is that reasonable?

lesser1 committed:
Ensure that token gets logged

Regardless of the environment.

lesser1 committed:
Change dialog height and width values to 500

Dialogs affected by this:
- AddTaxRateDialog
- AddTaxRateVersionDialog
- AddToWishlistDialog
- SwitchUserRoleDialog

lesser1 committed:
Add explanatory text to SwitchUserRoleDialog

Also, run "yarn lint ; yarn format"

lesser1 committed:
Remove unnecessary emitted event...

... from SwitchUserRoleDialog. Add boolean parameter to the 'close'
event. Modify code that handles the event in TheAppBar.vue.

lesser1 committed:
Refactoring:

    - Extract UserRole into separate file.
    - Extract extra functions out of initLogin in app.ts.
    - Add helper getters to the pinia store (app.ts).
    - Use storeToRef refs instead of local computed refs.
    - Rewrite various JSDoc comments.

lesser1 committed:
Rename user roles to match Keycloak role names

lesser1 committed:
Merge branch 'main' into feature/add-user-roles-to-app-store
